HipercowApi/Tools/ExceptionHandlingMiddleware.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Imperial College London. All rights reserved.
+
+namespace HipercowAPI.Tools
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Net;
+    using System.Text.Json;
+    using Hipercow_api.Tools.Exceptions;
+
+    /// <summary>
+    /// Class for exception middleware.
+    /// </summary>
+    /// <param name="next">Next delegate.</param>
+    /// <param name="logger">Logger.</param>
+    [ExcludeFromCodeCoverage]
+    public class ExceptionHandlingMiddleware(
+        RequestDelegate next,
+        ILogger<ExceptionHandlingMiddleware> logger)
+    {
+        private readonly RequestDelegate _next = next;
+        private readonly ILogger<ExceptionHandlingMiddleware> _logger = logger;
+
+        /// <summary>
+        /// Error invocation handler.
+        /// </summary>
+        /// <param name="context">Context.</param>
+        /// <returns>Task.</returns>
+        public async Task InvokeAsync(HttpContext context)
+        {
+            try
+            {
+                await _next(context);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "An unhandled exception occurred");
```

Comment:
We should be logging the error with a UID which we also return in the response (for 500s only) so we can look up the problem response in the log. 

What does this actually log? I think we want it to be all the error details including stack, so we can work out what the error was. MVC docs pretty opaque. Do you have an example of what actually gets logged. 

Also - this is getting logged for all errors, but I don't think we need full error details for anything other than a 500. We don't need a stack trace for 400, 401 or 403.

Reply:
I think it logs the "stack trace, message and inner exceptions". Think I've got the UID in now - will try and add testing to the middleware to see what output we get...